Add leak-tracked ref counting to MultiSearchResponse

original-brownbear opened the pull request:
Make the ref-counting leak tracked and release child items for MultiSearchResponse and adjust tests + REST action accordingly.

for #102030

original-brownbear committed:
Add leak-tracked ref counting to MultiSearchResponse

Make the ref-counting leak tracked and release child items for
MultiSearchResponse and adjust tests + REST action accordingly.

original-brownbear committed:
fix more tests

original-brownbear committed:
Fix ref counting when using futures in AbstractClient

We fundamentally need to increment the ref count by one before
assigning a ref counted value to a future result.
Otherwise, trasnsport actions will have to understand the specific
kind of listener they resolve and cannot decrement sent/consumed
transport messages themselves.

original-brownbear committed:
Merge branch 'fix-client-ref-counting' into ref-count-multi-search-response

original-brownbear committed:
fixes

original-brownbear committed:
Merge remote-tracking branch 'elastic/main' into ref-count-multi-search-response

original-brownbear committed:
test

original-brownbear committed:
Merge remote-tracking branch 'elastic/main' into ref-count-multi-search-response

original-brownbear commented:
WIP for now lets see what tests say first + this has a dependency on https://github.com/elastic/elasticsearch/pull/102607 which should be discussed first

original-brownbear committed:
fixes

original-brownbear committed:
Merge remote-tracking branch 'elastic/main' into ref-count-multi-search-response

original-brownbear committed:
fixes

original-brownbear committed:
Merge remote-tracking branch 'elastic/main' into ref-count-multi-search-response

original-brownbear committed:
fix single shard action

original-brownbear committed:
Merge remote-tracking branch 'elastic/main' into ref-count-multi-search-response

original-brownbear committed:
fix ref couted future

original-brownbear committed:
Merge remote-tracking branch 'elastic/main' into ref-count-multi-search-response

original-brownbear committed:
fix fleet

original-brownbear committed:
fix fleet

original-brownbear committed:
fix test

original-brownbear committed:
Merge remote-tracking branch 'elastic/main' into ref-count-multi-search-response

original-brownbear committed:
asserts

original-brownbear committed:
Merge remote-tracking branch 'elastic/main' into ref-count-multi-search-response

original-brownbear committed:
Merge remote-tracking branch 'elastic/main' into ref-count-multi-search-response

original-brownbear committed:
more assert

original-brownbear commented:
extracted the general transport fix to https://github.com/elastic/elasticsearch/pull/102696 lets review that first maybe :)

original-brownbear committed:
Merge remote-tracking branch 'elastic/main' into ref-count-multi-search-response

kpollich reviewed (approved):
Fleet codeowner 🚀

original-brownbear committed:
Merge remote-tracking branch 'elastic/main' into ref-count-multi-search-response

original-brownbear commented on modules/lang-mustache/src/main/java/org/elasticsearch/script/mustache/MultiSearchTemplateResponse.java:
```diff
@@ -170,18 +170,22 @@ static final class Fields {
     public static MultiSearchTemplateResponse fromXContext(XContentParser parser) {
         // The MultiSearchTemplateResponse is identical to the multi search response so we reuse the parsing logic in multi search response
         MultiSearchResponse mSearchResponse = MultiSearchResponse.fromXContext(parser);
```
No these aren't releasable as of right now. Not sure it's worth adding it either. Maybe we could add that interface to transport messages but that's a discussion for distrib and some other time I think :)

benwtrent commented on x-pack/plugin/enrich/src/main/java/org/elasticsearch/xpack/enrich/action/EnrichCoordinatorProxyAction.java:
```diff
@@ -239,12 +239,22 @@ static BiConsumer<MultiSearchRequest, BiConsumer<MultiSearchResponse, Exception>
                     ActionListener<MultiSearchResponse> listener = ActionListener.wrap(response -> {
                         shardResponses.put(enrichIndexName, new Tuple<>(response, null));
```
This is tricky to me, `shardResponses` is a map outside the listener and keeps a copy of the `MultiSearchResponse`. This happens `itemsPerIndex` times. Each time, the listener "exits" with the map keeping a reference.

Then once all the searches are completed, it is reduced, potentially creating a NEW multisearchresponse with references to the response items in the individually collected responses.

I wonder if `shardResponses` should increment the references when storing the values, and then once this is all completed, decrement them.

original-brownbear replied:
Hmm you're right there's a massive testing gap here in the enrich codebase. We only ever run with the counter going up to `1` so we just never hit this. It's definitely a bug, I'll address enhancing the tests first in a separate PR tomorrow and then we can get back to this, lets not mix this stuff :)

original-brownbear replied:
https://github.com/elastic/elasticsearch/pull/102751 once that's in we should see this issue reproduce here

original-brownbear replied:
Alright, I merged the above PR in here and it reproduced this issue. I fixed it just the way you describe above and it should be fine now.  If you're fine with this now and reviewing the other PR takes time, I can just back the test out of this PR before merging :)

original-brownbear committed:
Add simple stress test to EnrichMultiNodeIT

Add a simple test that forces enrich tasks to queue up to guard against
bugs around holding on to multi search respones in this logic, now that we
are introducing ref counting to them.

relates #102479

original-brownbear committed:
Merge remote-tracking branch 'elastic/main' into ref-count-multi-search-response

original-brownbear committed:
Merge remote-tracking branch 'origin/enhance-enrich-tests' into ref-count-multi-search-response

original-brownbear committed:
handle enrich reduce correctly

piergm reviewed (approved):
LGTM!

benwtrent commented on x-pack/plugin/enrich/src/main/java/org/elasticsearch/xpack/enrich/action/EnrichCoordinatorProxyAction.java:
```diff
@@ -238,13 +239,24 @@ static BiConsumer<MultiSearchRequest, BiConsumer<MultiSearchResponse, Exception>
                     final List<Tuple<Integer, SearchRequest>> enrichIndexRequestsAndSlots = entry.getValue();
                     ActionListener<MultiSearchResponse> listener = ActionListener.wrap(response -> {
                         shardResponses.put(enrichIndexName, new Tuple<>(response, null));
+                        response.incRef(); // will be released during
```
```suggestion
                        response.incRef(); // will be released during reduce
```

benwtrent reviewed (approved):
I think this looks good now. Shows Enrich needs more tests!

original-brownbear committed:
remove test and fix comment

original-brownbear committed:
Merge remote-tracking branch 'elastic/main' into ref-count-multi-search-response

original-brownbear commented:
Thanks all!
